x/qgb/abci.go

```diff
@@ -15,13 +15,15 @@ const SignificantPowerDifferenceThreshold = 0.05
 
 // EndBlocker is called at the end of every block.
 func EndBlocker(ctx sdk.Context, k keeper.Keeper) {
-	handleDataCommitmentRequest(ctx, k)
+	// we always want to create the valset at first so that if there is a new validator set, then it is
+	// the one responsible for signing from now on.
 	handleValsetRequest(ctx, k)
+	handleDataCommitmentRequest(ctx, k)
```

Comment:
Can you explain to me a little more the rationale for switching this? Is one dependent on the other?

Reply:
This is done mainly in the case of a hardfork where there is a manual change to the validator set. This would ensure that straight after starting the chain again, a new valset is created representing the new validator set, and that one will be used to sign the new data commitments.

Otherwise, if we're in a case where a new data commitment is created straight after the hardfork, we will be in the situation where that data commitment will need to be signed by the previous validator set which can not exist anymore. 

So, it's better to always check if we there is a change to the validator set and account for it, before creating any data commitment

Reply:
So the validator set of that height signs over the data hashes for the last n heights?  (as opposed to the validator set at the previous height)

Reply:
yes, in the event of a validator set change, we want to change directly to the new one and use it for signatures